[BACKEND] Implement 3xTF32 trick

lezcano opened the pull request:
This PR implements the [3xTF32 trick](https://github.com/NVIDIA/cutlass/discussions/385) to make use of the TCs on F32 tensors without sacrificing accuracy. This is particularly relevant for PyTorch, as TF32 is off by default.

Benchmarks on A100 from `python/tutorials/03-matrix-multiplication.py` run on `float32` data using `use_tf32=False`:
```
         M       N       K     cuBLAS     Triton    This PR
0    256.0   256.0   256.0   1.927529   1.092267   1.489455
1    384.0   384.0   384.0   5.026909   3.567484   3.686400
2    512.0   512.0   512.0   8.192000   6.553600   6.898527
3    640.0   640.0   640.0  12.190476  10.448980  10.666666
4    768.0   768.0   768.0  13.405091  10.287628  14.503869
5    896.0   896.0   896.0  14.049280  13.380267  20.070399
6   1024.0  1024.0  1024.0  15.887515  12.264046  19.239927
7   1152.0  1152.0  1152.0  16.681475  15.633424  24.883201
8   1280.0  1280.0  1280.0  16.516129  15.340824  28.248276
9   1408.0  1408.0  1408.0  17.090206  14.774461  24.016635
10  1536.0  1536.0  1536.0  17.014154  15.624477  26.021647
11  1664.0  1664.0  1664.0  17.043394  15.073554  25.858942
12  1792.0  1792.0  1792.0  17.107190  16.171833  29.577431
13  1920.0  1920.0  1920.0  17.883570  15.762828  26.331430
14  2048.0  2048.0  2048.0  17.623127  17.032706  27.413751
15  2176.0  2176.0  2176.0  17.887688  16.686275  29.945905
16  2304.0  2304.0  2304.0  19.019006  17.933838  33.787654
17  2432.0  2432.0  2432.0  17.940270  17.288901  31.181425
18  2560.0  2560.0  2560.0  18.164080  17.075561  31.844508
19  2688.0  2688.0  2688.0  17.594183  16.703239  30.370742
20  2816.0  2816.0  2816.0  18.766871  18.089676  33.242537
21  2944.0  2944.0  2944.0  18.735350  17.855977  33.695763
22  3072.0  3072.0  3072.0  18.420008  17.766898  32.768000
23  3200.0  3200.0  3200.0  18.470418  17.704011  33.255391
24  3328.0  3328.0  3328.0  18.253370  17.710036  32.753092
25  3456.0  3456.0  3456.0  18.546485  17.793328  33.634362
26  3584.0  3584.0  3584.0  18.368824  17.833278  33.142423
27  3712.0  3712.0  3712.0  18.665424  17.938112  34.036574
28  3840.0  3840.0  3840.0  18.638578  18.076496  33.794348
29  3968.0  3968.0  3968.0  18.965486  18.190808  34.324595
30  4096.0  4096.0  4096.0  19.035276  18.365864  34.450135
```

It's an overall win,  getting roughly a 85% speed-up on large sizes.

Note that the rounding is differs a little bit to the one [implemented in CUTLASS](https://github.com/NVIDIA/cutlass/blob/a8f2c80db0564c74f4efccac71993b971dfc448b/include/cutlass/gemm/warp/mma_tensor_op_fast_f32.h#L99-L100). We could implement that rounding if we wanted though.

This is still a bit far from the 2x speed-ups announced by CUTLASS. To get close to those numbers, we should probably need to remove the stores to shared before `ldmatrix`.

jlebar commented on python/triton/language/semantic.py:
```diff
@@ -1347,6 +1348,28 @@ def assert_dtypes_valid(lhs_dtype, rhs_dtype, options):
         ret_scalar_ty = out_dtype
 
+    if not allow_tf32 and lhs.dtype.is_fp32() and rhs.dtype.is_fp32() and getattr(builder.options, "allow_tf32"):
+        # Implement 3xTF32 trick
```
Could we add a link here in the source code?

jlebar commented on third_party/nvidia/backend/compiler.py:
```diff
@@ -62,6 +62,7 @@ class CUDAOptions:
     ptx_version: int = None
     enable_fp_fusion: bool = True
     allow_fp8e4nv: bool = False
+    allow_tf32: bool = False
```
I'm not sure about the name of this option.  Without the context in this patch, I'd think this means, do we allow implicitly upgrading fp32 dots to tf32, which is not what this controls.

Perhaps supports_tf32, with a comment saying that this indicates whether the hw supports tf32 but doesn't give us permission to *use* tf32 to silently, except where the result is "as if" we'd computed in fp32?

lezcano replied:
Yeah, I wasn't sure about this name. I followed the convention of `allow_fp8e4nv`, but I agree that `supports_tf32` is a much better name. Let me change it.

jlebar reviewed (commented):
This is awesome.

jlebar reviewed (approved):
Thank you!

Jokeren commented on python/triton/language/semantic.py:
```diff
@@ -1347,6 +1348,24 @@ def assert_dtypes_valid(lhs_dtype, rhs_dtype, options):
         _0 = builder.get_fp16(0) if out_dtype.is_fp16() else builder.get_fp32(0)
         ret_scalar_ty = out_dtype
 
+    if not allow_tf32 and lhs.dtype.is_fp32() and rhs.dtype.is_fp32() and getattr(builder.options, "supports_tf32"):
```
What if the builder doesn't have a `supports_tf32` option?

lezcano replied:
getattr will return None and it'll evaluate to False.

ThomasRaoux commented on python/triton/language/semantic.py:
```diff
@@ -1347,6 +1348,24 @@ def assert_dtypes_valid(lhs_dtype, rhs_dtype, options):
+        # Implement 3xTF32 trick
+        # https://github.com/NVIDIA/cutlass/discussions/385
+
+        def fp32_to_tf32(x):
+            return tl.inline_asm_elementwise(asm="""cvt.rna.tf32.f32 $0, $2;\n\tsub.f32 $1, $2, $0;""",
+                                             constraints="=r,=r,r", args=[x], dtype=[tl.float32, tl.float32],
+                                             is_pure=True, pack=1, _builder=builder)
+
+        xb, xs = fp32_to_tf32(lhs)
+        yb, ys = fp32_to_tf32(rhs)
+        del kwargs["lhs"]
+        del kwargs["rhs"]
+        del kwargs["allow_tf32"]
+        kwargs["acc"] = dot(xb, ys, allow_tf32=True, **kwargs)
+        kwargs["acc"] = dot(xs, yb, allow_tf32=True, **kwargs)
+        return dot(xb, yb, allow_tf32=True, **kwargs)
+
```
This is cool however my understanding is that the 3xTF32 precision doesn't match exactly fp32 precision which will be a problem for user who want strict precision.

lezcano replied:
Its precision is better than fp32 so that should be fine?

ThomasRaoux replied:
I wonder if we should have a separate control for this

lezcano replied:
At least that's fine in PyTorch. Our policy is that `torch.compile` should have the same accuracy or better than eager mode.

ThomasRaoux replied:
> Its precision is better than fp32 so that should be fine?

I don't think so, I believe we want it to match

ThomasRaoux replied:
> At least that's fine in PyTorch. Our policy is that torch.compile should have the same accuracy or better than eager mode.

We have users that rely on allow_tf32=False to have a precision closer to the CPU

ThomasRaoux replied:
I wonder if this is something that can be done on user side? For instance inductor could implement it with a function wrapper?

jlebar replied:
Personally I actually really like this being in the compiler, because it's our speciality to figure out the fastest way to run your code (within whatever correctness constraints we set, of course).  I don't think we want every user to have to re-implement this trick.

ThomasRaoux replied:
it could be part of standard but I don't think we should change the precision under the hood

jlebar replied:
No argument against your point that we should not change the precision of the dot without user consent.

ThomasRaoux reviewed (changes requested):
I think there is going to be a precision problem. Maybe we would need a separate control for it but we want the precision to match fp32

lezcano commented:
A few updates:
- After looking a bit more closely at the code with @alexsamardzic and @peterbell10, we found that we were able to squeeze some extra performance going up to 33TFPS. To do this, we need to implement this patch after the `c += tl.dot(a,b) -> tl.dot(a,b,c)` transformation has been done. This performance is pretty much 1/3 of that of tf32, which makes sense. Pipelining should take us close to the 40TFPS announced by CUTLASS
- It is not kosher to implement optimisations in the frontend. Even less so those that are backend-specific.
- For these two reasons, it makes much more sense to write this optimisation as a TTGIR optimisation pass. I hope to be able to put a PR tomorrow.

As for how to expose this. I agree that it'd be good to give the user the possibility to use the IEEE version. I can either add a new `f32_backend: str` flag (happy to bikeshed the name) exposing this. Now, I really think that even if the user can use the IEEE version, the default behaviour should be to use the 3xtf32 version. We already default to `use_tf32=true` for speed disregarding accuracy. I think we should also default to speed (but even accuracy!) in this case.

ThomasRaoux commented:
> after the c += tl.dot(a,b) -> tl.dot(a,b,c) transformation has been done. 

`tl.dot` supports 3 arguments so you should be able to write it directly in the language. Would that allow you to do it as a library? 

> It is not kosher to implement optimisations in the frontend. Even less so those that are backend-specific.

It depends if we consider it to be a library level or a backend optimization.

> For these two reasons, it makes much more sense to write this optimisation as a TTGIR optimisation pass. I hope to be able to put a PR tomorrow.

The downside of this strategy is that we need to add another attribute to language. As @sbodenstein pointed out there may be a bunch of other algorithm and we don't want to keep adding attributes to the language.


> As for how to expose this. I agree that it'd be good to give the user the possibility to use the IEEE version. I can either add a new f32_backend: str flag (happy to bikeshed the name) exposing this. Now, I really think that even if the user can use the IEEE version, the default behaviour should be to use the 3xtf32 version. We already default to use_tf32=true for speed disregarding accuracy. I think we should also default to speed (but even accuracy!) in this case.

The current default behavior is tf32, are you suggesting changing it to 3xtf32? Or are you saying it should be default if use_tf32=False? In this case default behavior doesn't mean much and the best would be having an enum of precisions.

I would still think a library solution if possible would be nice.

Jokeren commented:
> For these two reasons, it makes much more sense to write this optimisation as a TTGIR optimisation pass. I hope to be able to put a PR tomorrow.

I have a different thought regarding this. I'm curious about what optimizations you'd like to implement? 

Maybe we can do this:

1. The backend can integrate optimizations for cases in which multiple dots exist in the loop (if I understand it correct) similar to your case.
2. The 3XTF32 or 4xTF32 solution can be a library API.

jlebar commented:
> tl.dot supports 3 arguments so you should be able to write it directly in the language. Would that allow you to do it as a library?

I think the concern is that if the user writes `c += core.dot(a, b)`, we need to first convert this into `tl.dot(a, b, c)` before we run the 3xtf32 optimization.

You *could* write the user code differently, but I think he's trying to make "reasonable user code" do the right thing.  Which, given our goal of portability, makes sense to me.

> The downside of this strategy is that we need to add another attribute to language.

We already have a tf32 attribute.  It actually sounds like he wants to *generalize* this into `f32_backend: str`, rather than necessarily adding a new attribute.  The generalized attribute would allow us to say that a dot runs in IEEE-fp32, tf32, 3xtf32, 4xtf32, or whatever other formats a backend supports.

OTOH if we don't do this, then it seems to me that every hardware vendor with its different dot precisions is going to want attributes on dot similar to the use_tf32 that we have today.

IOW it feels like this is a *solution* to the problem of attribute-creep?

ptillet commented:
I personally think that replacing `allow_tf32=True` with an `fp32_backend` flag is a good change. It is more future-proof, as if TF32 stops existing (or the specs change) the `allow_tf32` attribute will become awkward to maintain. Not to mention that TF32 is an nvidia-specific format, so even today it is weird to specify `allow_tf32=True` on non-nvidia hardware. 

However, I do think the default for `fp32_backend` should be `target.default_fp32_backend`, which should be `tf32` on nvidia. And of course each target should also have a `supported_fp32_backend` attribute so that users get an error if they try to use `tf32` on AMD GPUs.

lezcano commented:
I completely agree with Philippe's comment. I'll go ahead and implement that.
Yes, I meant the default behaviour for `use_tf32=False`. I do think that the default behaviour should be to do the matmul in TF32, in line with the rest of the language (cf. `tl.sin`). If we do the BC breaking change `use_tf32` -> `f32_backend`, then this point is moot. Now, this point is still relevant if we want to keep both in the API for BC compat reasons. For now, I'll implement it in a BC compat way mapping `use_tf32=False` -> `f32_backend="3xtf32"` and throwing a deprecating warning, but I'm happy to make the BC breaking change if that's alright. Whatever you guys prefer cc @ptillet

lezcano commented:
Updated the PR. In particular:
- This transformation is implemented as a TTGIR pass
- Changed `allow_tf32: bool` into `f32_backend`. Removed `allow_tf32` (bc-breaking) as otherwise the semantics with both args present are rather unclear.
- Documented this flag + added per-device defaults for NVIDIA and AMD
- Updated the speed-ups in the OP: We have a 85% speed-up.

The PR may be easier to review commit-by-commit, given the amount of files it touches.

ThomasRaoux commented on lib/Dialect/TritonGPU/Transforms/FP32DotTC.cpp:
```diff
@@ -0,0 +1,93 @@
+#define GEN_PASS_CLASSES
+#include "triton/Dialect/TritonGPU/Transforms/Passes.h.inc"
+
+struct FP32DotTCPass : public TritonGPUFP32DotTCBase<FP32DotTCPass> {
+  void runOnOperation() override {
+    MLIRContext *context = &getContext();
+    ModuleOp m = getOperation();
+
+    mlir::RewritePatternSet decomposePatterns(context);
+    decomposePatterns.add<TF32x3>(context);
+    if (mlir::applyPatternsAndFoldGreedily(m, std::move(decomposePatterns))
+            .failed()) {
+      signalPassFailure();
+    }
+  }
+};
```
nit: should also go in anonymous namespace

jlebar replied:
I don't think that's correct.  The passes header declares all the pass classes, and those classes need to be in a consistent namespace.  Right not we put them in the global namespace, though it would be better to put them in namespace mlir::triton.  In any case they can't be in the anon ns.

ThomasRaoux replied:
the base class is declared in the header, the implementation class can be in an anonymous namespace. We definitely have other passes done that way.

ThomasRaoux commented on include/triton/Dialect/Triton/IR/TritonOps.td:
```diff
@@ -565,7 +565,7 @@ def TT_DotOp : TT_Op<"dot", [Pure,
       TT_FpIntTensor:$b,
       TT_FpIntTensor:$c,
-      BoolAttr:$allowTF32,
+      StrAttr:$f32Backend,
```
nit: making it an enum would be a bit less error-prone. You can use `I32EnumAttr` for that

jlebar replied:
If we use an enum, do we have to list all possible values in TritonOps.td?  That doesn't work well for out-of-tree backends.

(I think it would be good to have the list of valid strings written down somewhere in the nvidia backend, though.)

ThomasRaoux replied:
correct the enum has to be in TritonOps.td, it still feels better than doing string checks? I believe we could decouple it using interfaces but that's probably an overkill to do it right now. Even if we have a lit of valid strings, it feels easy to make a typo when checkin the value. 
I'd be fine with the possible constants be declared in a header as constexpr and used instead of literal strings

jlebar replied:
> correct the enum has to be in TritonOps.td, it still feels better than doing string checks?

I just thought we had it as a goal to make it possible for people to write out-of-tree backends?  If you think that's not relevant here, then definitely enums would be better.

> I'd be fine with the possible constants be declared in a header as constexpr and used instead of literal strings

The problem is there's no way to enforce this, and people will use strings anyway.  (And indeed the IR uses a string, so that's the place where we're most likely to have a typo.)  So I think there's an argument for leaning into the strings in the C++ and accepting that we need to write tests.  At least, that is my experience with this sort of thing from XLA, where it's used a lot.

Again I do think it needs to be documented in the backend which strings are acceptable.  Right now I only see it in the frontend.  I also think (orthogonal to all this) that we need a check during lowering that we know how to lower the string we got (i.e. check for invalid strings during lowering -- I don't see this anywhere) in order to catch typos in the IR.

lezcano replied:
I went with the enum implementation, as we already use enums in quite a few places. Going from enums to any other representation in the future should be trivial tho.

ThomasRaoux reviewed (commented):
That looks good, added couple minor comments

jlebar commented on include/triton/Dialect/Triton/IR/TritonOps.td:
```diff
@@ -565,7 +565,7 @@ def TT_DotOp : TT_Op<"dot", [Pure,
       TT_FpIntTensor:$b,
       TT_FpIntTensor:$c,
-      BoolAttr:$allowTF32,
+      StrAttr:$f32Backend,
```
Please write documentation explaining what this field does.

jlebar commented on include/triton/Dialect/Triton/IR/TritonOps.td:
```diff
@@ -565,7 +565,7 @@ def TT_DotOp : TT_Op<"dot", [Pure,
       TT_FpIntTensor:$b,
       TT_FpIntTensor:$c,
-      BoolAttr:$allowTF32,
+      StrAttr:$f32Backend,
```
Perhaps "f32Backend" isn't the best name.  We can use this field for the f16 (or whatever) backend too, right?  No need to have multiple fields depending on the dtype.

lezcano replied:
I just followed https://github.com/openai/triton/pull/3234#issuecomment-1971751206. I'm happy to bikeshed if a better name / convention is proposed.

jlebar replied:
`$precision` seems good to me.

jlebar commented on lib/Dialect/TritonGPU/Transforms/FP32DotTC.cpp:
```diff
@@ -0,0 +1,93 @@
+
+  LogicalResult
+  matchAndRewrite(tt::DotOp dotOp,
+                  mlir::PatternRewriter &rewriter) const override {
```
We are `using namespace mlir`, so we don't need to qualify `mlir::` here and elsewhere.

jlebar commented on lib/Dialect/TritonGPU/Transforms/FP32DotTC.cpp:
```diff
@@ -0,0 +1,93 @@
+          .isF32();
+    };
+
+    if (!(dotOp.getF32Backend() == "3xtf32" && isF32(dotOp.getOperand(0)) &&
```
You can do dotOp.getA/B/C().  These are also always RankedTensorType'ed values, so you can leave off a cast in some places.

jlebar commented on lib/Dialect/TritonGPU/Transforms/FP32DotTC.cpp:
```diff
@@ -0,0 +1,93 @@
+          .create<tt::ElementwiseInlineAsmOp>(
+              dotOp.getLoc(), value.getType(), "cvt.rna.tf32.f32 $0, $1;",
+              "=r,r",
+              /*isPure*/ true, /*pack*/ 1, std::vector<mlir::Value>{value})
```
Please write as /*isPure=*/true, /*pack=*/1.  That format can be checked by clang-tidy.

lezcano replied:
we'll have AGI before we have decent parsers :P

jlebar commented on lib/Dialect/TritonGPU/Transforms/FP32DotTC.cpp:
```diff
@@ -0,0 +1,93 @@
+          .create<tt::ElementwiseInlineAsmOp>(
+              dotOp.getLoc(), value.getType(), "cvt.rna.tf32.f32 $0, $1;",
+              "=r,r",
+              /*isPure*/ true, /*pack*/ 1, std::vector<mlir::Value>{value})
```
Nit, I don't think you need to construct a vector -- an ArrayRef or, if necessary SmallVector, should work?

jlebar commented on lib/Dialect/TritonGPU/Transforms/FP32DotTC.cpp:
```diff
@@ -0,0 +1,93 @@
+              dotOp.getLoc(), value.getType(), "cvt.rna.tf32.f32 $0, $1;",
+              "=r,r",
+              /*isPure*/ true, /*pack*/ 1, std::vector<mlir::Value>{value})
+          .getResult()[0];
```
Does getResult(0) work?

lezcano replied:
Nope
```
/home/lezcano/git/triton/python/build/cmake.linux-x86_64-cpython-3.8/include/triton/Dialect/Triton/IR/Ops.h.inc:2568:35: note: candidate: 'mlir::Operation::result_range mlir::triton::ElementwiseInlineAsmOp::getResult()'
 2568 |   ::mlir::Operation::result_range getResult();
      |                                   ^~~~~~~~~
/home/lezcano/git/triton/python/build/cmake.linux-x86_64-cpython-3.8/include/triton/Dialect/Triton/IR/Ops.h.inc:2568:35: note:   candidate expects 0 arguments, 1 provided
```

jlebar commented on lib/Dialect/TritonGPU/Transforms/FP32DotTC.cpp:
```diff
@@ -0,0 +1,93 @@
+                                        "tf32", dotOp.getMaxNumImpreciseAcc());
+    };
+    auto a = dotOp.getOperand(0);
+    auto b = dotOp.getOperand(1);
```
dotOp.getA/B().  TBH I'd just remove the variables entirely, but up to you.

jlebar commented on lib/Dialect/TritonGPU/Transforms/FP32DotTC.cpp:
```diff
@@ -0,0 +1,93 @@
+    auto bBig = f32ToTF32(b);
+    auto bSmall = sub(b, bBig);
+
+    auto dot1 = dot(aSmall, bBig, dotOp.getOperand(2));
```
dotOp.getC()

jlebar commented on third_party/nvidia/backend/compiler.py:
```diff
@@ -62,6 +62,8 @@ class CUDAOptions:
     enable_fp_fusion: bool = True
     allow_fp8e4nv: bool = False
+    default_f32_backend: str = "tf32"
+    allowed_f32_backends: Tuple[str] = ("tf32", "3xtf32", "ieee")
```
These properties should have the word "dot" in them somewhere.

jlebar commented on third_party/nvidia/backend/compiler.py:
```diff
@@ -128,6 +130,8 @@ def make_ttgir(mod, metadata, opt, capability):
         # optimize TTGIR
         passes.ttgpuir.add_coalesce(pm)
+        if capability // 10 >= 8:
+            passes.ttgpuir.add_fp32_dot_tc(pm)
```
What happens if I pass "3xtf32" but I'm not on `capability // 10 >= 8`?

lezcano replied:
We fallback transparently to `ieee`, same as we do now with `allow_tf32=True` now. Added to the docs in TritonOps.td and the public docs.

jlebar replied:
Also Phil (who's the boss here) said

> And of course each target should also have a supported_fp32_backend attribute so that users get an error if they try to use tf32 on AMD GPUs.

so it seems to me like one could extend this logic to different nvidia GPUs just as easily as to nvidia vs AMD.

lezcano replied:
Note that we already "transparently downgrade" `allow_tf32=True` to `allow_tf32=False` if there are no TC.

About `"foobar"`, that'll throw when we use enums. I'll push that implementation soon.

lezcano commented:
Addressed the reviews. Should we bikeshed the name tho? @jlebar proposes "precision" for it to be more generic.

ThomasRaoux commented:
> Addressed the reviews. Should we bikeshed the name tho? @jlebar proposes "precision" for it to be more generic.

The name what proposed by @ptillet who is in vacation until Monday. Not sure we can discuss it without him.

Other than the name where I don't have a strong opinion I think this looks good. 

Note that I did add an env var override for TFF32 behavior: https://github.com/openai/triton/pull/3290 which might be in your way. If you want to rebase with it that's great but feel free to break it and I'll add it back after your change

ThomasRaoux reviewed (approved):
That looks good to me modulo the naming decision

jlebar commented:
If you're not comfortable making an executive decision, then I propose we use the current name and I volunteer as tribute to run `sed` if I can convince Phil that `precision` is a better name.

jlebar commented on include/triton/Dialect/Triton/IR/TritonOps.td:
```diff
@@ -560,14 +560,19 @@ def TT_DotOp : TT_Op<"dot", [Pure,
+        tf32: use TC with tf32 ops.
+        tf32x3: implement the 3xTF32 trick. For more info see the pass in F32DotTC.cpp
+        ieee: don't use TC, implement dot in software.
+        If the GPU does not have Tensor cores or the inputs are not f32, this flag is ignored.
```
Presumably these descriptions should be on the enum now.

jlebar replied:
I think this comment still applies?

lezcano replied:
I figured it'd be better to leave them in the place where one would first look at when finding what this flag really does, but sure, I can move them

jlebar replied:
For example the enum is also used on dot_async.

(If you wanted a "see X" comment I'd be fine with that, although it's probably pretty obvious where to look.)

jlebar commented on test/Analysis/test-alias.mlir:
```diff
@@ -29,7 +29,7 @@ tt.func @matmul_loop(%lb : index, %ub : index, %step : index, %A : !tt.ptr<f16>,
     %b_ = tt.load %b_ptr, %b_mask, %b_other {cache = 1 : i32, evict = 1 : i32, isVolatile = false} : tensor<32x128xf16, #BL>
     %b = triton_gpu.convert_layout %b_ : tensor<32x128xf16, #BL> -> tensor<32x128xf16, #B_DOT>
-    %c = tt.dot %a, %b, %prev_c {transA = false, transB = false, allowTF32 = true, maxNumImpreciseAcc = 0 : i32} : tensor<128x32xf16, #A_DOT> * tensor<32x128xf16, #B_DOT> -> tensor<128x128xf32, #C>
+    %c = tt.dot %a, %b, %prev_c {transA = false, transB = false, f32Backend = 0 : i32, maxNumImpreciseAcc = 0 : i32} : tensor<128x32xf16, #A_DOT> * tensor<32x128xf16, #B_DOT> -> tensor<128x128xf32, #C>
```
Is it really not possible to spell out the enumerator here, we have to use opaque numbers in the IR??

ThomasRaoux replied:
MLIR does support printing the associated string, there is an example here:
https://github.com/llvm/llvm-project/blob/8b5b294ec2cf876bc5eb5bd5fcb56ef487e36d60/mlir/include/mlir/Dialect/GPU/IR/GPUOps.td#L40

lezcano replied:
I tried to put the name the for quite some time before putting up the PR with little success. I think the issue is that in https://github.com/openai/triton/blob/dd8c270e5271523b9672eaa515117e6f2f5f0be6/include/triton/Dialect/Triton/IR/CMakeLists.txt#L20-L21
we are not putting these in a dialect. Then, according to [this post](https://discourse.llvm.org/t/assembly-format-of-an-array-of-enumattr/64936), we would need to refer to them as `#dialect<enum tf32>`. I tried exporting them as part of a dialect, but it broke other things.

I would like some pointers on how to do it, as I'm not fluent on MLIR.

jlebar replied:
@ThomasRaoux do you have cycles to help out with this?  If not IMO we should land it as a string, because having magic numbers like `0` in the IR is no bueno.  (I'm willing to help, but you know a lot more about MLIR than I do...)

jlebar replied:
Sounds like Thomas does not have time to help out with this right now.  We're OK merging as-is and figuring out how to fix this as a follow-up.

lezcano commented:
I'll rebase tomorrow and address the comments. The idea of landing + forward fixing the naming at a later stage sgtm.

lezcano commented:
Rebased. This should be ready modulo CI agreeing + naming the name of the kwarg + changing the assembly of the property. These two can be easily implemented later via a search/replace as forward fixes.

lezcano commented:
This should be ready to run CI

lezcano commented:
I missed that one in the last rebase. Fixed.

ptillet commented:
Back from vacation! I think `precision` is a better name, although `input_precision` would probably be even better since it would be clear that it does not apply to the internal accumulators

jlebar commented:
> input_precision would probably be even better since it would be clear that it does not apply to the internal accumulators

Doesn't the 3xtf32 trick apply to accumulators?  (I mean I guess it's sort of both the inputs and the accumulators, in a way?)

@ptillet do you have a preference on enum vs string for this field?  String is going to work better with out-of-tree backends, but if we don't care about that then enum is definitely better.  Although if we go with enum we also need to figure out how to get the enumeration string to print in the IR.

lezcano commented on python/triton/language/semantic.py:
```diff
@@ -1289,7 +1289,16 @@ def atomic_xchg(ptr: tl.tensor, val: tl.tensor, mask: tl.tensor, sem: str, scope
 
-def dot(lhs: tl.tensor, rhs: tl.tensor, acc: tl.tensor, allow_tf32: bool, max_num_imprecise_acc: int,
+def _str_to_dot_f32_backend(f32_backend, builder):
+    assert f32_backend.lower(
+    ) in builder.options.allowed_dot_f32_backends, f"f32 _backend must be one of {builder.options.allowed_dot_f32_backends}. Got {f32_backend}"
```
yapf insists we should format this long line like this... lol

lezcano replied:
I just used a `\`. I opened an issue at https://github.com/google/yapf/issues/1207, but the whole yapf repo doesn't look particularly active...

lezcano commented:
Between precision vs. input_precision, I like precision better, as it's more concise.

ptillet commented:
> Doesn't the 3xtf32 trick apply to accumulators? (I mean I guess it's sort of both the inputs and the accumulators, in a way?)

Well, the goal is to improve the precision of the input operands, isn't it? The accumulator is always FP32, regardless of whether the tensor cores are operating in TF32 mode or not. 

Besides, we have separate flags to control the precision of the output/accumulator (`out_dtype` and `max_num_imprecise_acc`)

> do you have a preference on enum vs string for this field?

I have a slight preference for strings here because this flag is almost always gonna be target-dependent

ThomasRaoux commented:
> > do you have a preference on enum vs string for this field?
> 
> I have a slight preference for strings here because this flag is almost always gonna be target-dependent

Note that even with enum we can support downstream targets by making an interface, that would be the cleanest solution in my opinion (although we could delay the interface part to later and start with what is there now)

ThomasRaoux commented:
> > Note that even with enum we can support downstream targets by making an interface
> 
> That sounds really interesting, do you have a code pointer for how that would work? (I'm unsure how it's supposed to work because an enum is internally an integer, so how does an out-of-tree backend "reserve" a numeric value?)

I guess it would have to be a per target enum at which point it is pretty much like a string, the only advantage is that we get the verifier for free and we don't manipulate strings explicitly. If we have verifier and create the strings either in the td file or as constexpr somewhere then we get the same advantages so this is as good of a solution.

lezcano commented:
So strings or enums? And what name do you want me to put?
Otherwise, could we get this one merge and then you guys can put up a follow-up with the bikeshedding? Whichever works best really.

ThomasRaoux commented:
> So strings or enums? And what name do you want me to put? Otherwise, could we get this one merge and then you guys can put up a follow-up with the bikeshedding? Whichever works best really.

I'm fine with either strings or enums, as it is only a backend change I think starting with either is fine as it is easy to change without user impact. 

Ideally we don't want to break bw compatibility multiple time so we should stick to a name. Since Phil prefers `input_precision` I think we should go with that.

lezcano commented:
Changed all the frontend keywords to `input_precision` and fixed a few hundred merge conflicts. Can you trigger CI see if I missed anything?

ThomasRaoux commented:
> Changed all the frontend keywords to `input_precision` and fixed a few hundred merge conflicts. Can you trigger CI see if I missed anything?

looks like you need to run pre-commit

jlebar commented:
Do `pre-commit install` and then you never need to remember to run pre-commit again.  :)

lezcano commented:
I have it installed! I guess it got as confused as I did with all these rebases...
I think it just skips files when doing an amend / rebase?

lezcano commented:
@ThomasRaoux @jlebar rebased

jlebar commented on include/triton/Dialect/Triton/IR/TritonOps.td:
```diff
@@ -566,14 +566,19 @@ def TT_DotOp : TT_Op<"dot", [Pure,
       TT_TensorOrMemDesc:$b,
       TT_FpIntTensor:$c,
-      BoolAttr:$allowTF32,
+      TT_F32BackendAttr:$f32Backend,
```
I think this should match the name of the Python attribute

ThomasRaoux reviewed (approved):
LGTM, let's wait @jlebar to approve as well

jlebar commented on python/test/unit/language/test_core.py:
```diff
@@ -2894,10 +2894,10 @@ def kernel(X, stride_xm, stride_xk, Y, stride_yk, stride_yn, W, stride_wn, strid
-                         DO_SOFTMAX=epilogue == 'softmax', CHAIN_DOT=epilogue == 'chain-dot', ALLOW_TF32=allow_tf32,
-                         num_warps=num_warps, num_ctas=num_ctas, out_dtype=out_dtype)
+                         DO_SOFTMAX=epilogue == 'softmax', CHAIN_DOT=epilogue == 'chain-dot',
+                         F32_BACKEND=input_precision, num_warps=num_warps, num_ctas=num_ctas, out_dtype=out_dtype)
```
rename F32_BACKEND

jlebar commented on python/src/ir.cc:
```diff
@@ -191,6 +191,12 @@ void init_triton_ir(py::module &&m) {
       .value("NONE", PropagateNan::NONE)
       .value("ALL", PropagateNan::ALL);
 
+  py::enum_<F32Backend>(m, "F32BACKEND", py::module_local())
```
Rename?

jlebar commented on python/triton/language/core.py:
```diff
@@ -1473,16 +1473,16 @@ def dot(input, other, acc=None, allow_tf32=None, max_num_imprecise_acc=None, out
     :type other: 2D tensor of scalar-type in {:code:`float16`, :code:`bfloat16`, :code:`float32`}
+    :param input_precision: How to exercise the Tenors cores for f32 x f32. If the device does not have Tensor Cores
+        or the inputs are not of dtype f32, this option is ignored.
+    :type other: string. Available options for nvidia: :code:`"tf32"`, :code:`"tf32x3"`, :code:`"ieee"`. Default: :code:`"tf32"`. Avaliable options for amd: :code:`"ieee"`.
```
This is the one place where we describe in the user documentation what these fields mean.  Seems like we should write a lot more and cite sources for further reading.

jlebar commented on python/triton/language/semantic.py:
```diff
@@ -1292,7 +1292,16 @@ def atomic_xchg(ptr: tl.tensor, val: tl.tensor, mask: tl.tensor, sem: str, scope
 
 
-def dot(lhs: tl.tensor, rhs: tl.tensor, acc: tl.tensor, allow_tf32: bool, max_num_imprecise_acc: int,
+def _str_to_dot_f32_backend(input_precision, builder):
```
rename

jlebar reviewed (changes requested):
Looks generally good to me, but I think we need to rename f32_backend to input_precision everywhere.  That and a few other smallish changes.

OAI is going on vacation next week and there's some ongoing discussion about whether we want to merge this on the Friday before everyone is out or if that's a Bad Idea.  In any case if we decide this should wait there shouldn't be as many merge conflicts as usual.

jlebar commented:
(We don't have a full story on bw compatibility, but for us, having a week or two where the old and new APIs are usable makes integration easier.  Because the next release will be 3.0, we're ok making bw-incompat changes in general.)

lezcano commented:
Alas, I'm leaving now (European time) and I don't have time to do those changes. If anyone wants to champion this through the finish line, feel free to push to my branch to get it merged.

jlebar commented:
Sorry we've been so slow on this one.  I'll probably be working some next week and I have approval from folks to merge this.  You have been very patient and I don't want to keep you in rebase hell indefinitely.

lezcano commented:
fwiw, I'll be on PTO on Monday / Tuesday, will be back on Wed. If that's too late for you, we can work on mergin this the Monday after. No rush.

lezcano commented:
@jlebar https://github.com/openai/triton/pull/3234/#discussion_r1526563951 and the previous comment still hold. Will add those tomorrow.
Other than that, I rebased, fixed the `lit` issues and renamed everything to "Input Precision". Would you mind running the CI? I'll fix the comments and anything that comes up in CI tomorrow first thing.

jlebar commented:
Just kicked CI for you.

Thanks again for pushing through with this one.
